### PR TITLE
Enabled media in ckeditor

### PR DESCRIPTION
**Pull request by marcelmoosbrugger**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | none
| Related issues/PRs | https://github.com/massiveart/husky/pull/658
| License | MIT
| Documentation PR | none

### Whats in this PR?
This is just a rebase and correction of: https://github.com/sulu/sulu/pull/2442

### CODE REVIEW

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/ckeditor/internal-link.js [hunk 1]
```diff
@@ -44,29 +45,74 @@ define([
...
+
+            getInternalLinkCommand: function(editor) {
+                return {
                     dialogName: 'internalLinkDialog',
                     allowedContent: 'sulu:link[title,target,sulu:validation-state,!href]',
                     requiredContent: 'sulu:link[href]',
                     exec: function() {
                         var $element = $('<div/>');
-                        $('body').append($element);
+                        $('#content').append($element);
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/ckeditor/internal-link.js]: Why is this change necessary?

File: src/Sulu/Bundle/MediaBundle/Controller/MediaController.php [hunk 2]
```diff
@@ -113,29 +113,35 @@ function ($id) use ($locale, $mediaManager) {
...
-                $listResponse[$i]['version'],
-                $listResponse[$i]['mimeType']
-            );
-            if (0 < count($format)) {
-                $listResponse[$i]['thumbnails'] = $format;
-            }
+        $listResponse = [];
+        $count = 0;
+
+        if ($request->get('ids') === null || count($ids) > 0) {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Controller/MediaController.php]: Shouldn't the first check be on `$ids`?

File: src/Sulu/Bundle/ContentBundle/Resources/public/js/ckeditor/internal-link.js (hunk 1 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/ContentBundle/Resources/public/js/ckeditor/internal-link.js]: Starting sandboxes outside the parents container or the body can lead to undesired bugs, because the component probably never gets destroyed. This was the case before changing this line. If you revert the change and open and close some internal link overlays in the ckeditor, you can see empty divs accumulating at the end of the body.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/ckeditor/media-link.js [hunk 3]
```diff
@@ -0,0 +1,190 @@
...
+                    icon: '/bundles/sulumedia/img/icon_remove_link_media.png',
+                    command: 'removeMediaLink',
+                    group: 'suluGroup'
+                });
+            },
+
+            onLoad: function() {
+                CKEDITOR.addCss(_.template(editorCSS, {
+                    tag: 'media',
+                    translations: {
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/ckeditor/media-link.js]: are the translations handled in the css file?

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection-overlay/main.js [hunk 4]
```diff
@@ -12,6 +12,8 @@
  *
  * @class MediaSelection
  * @constructor
+ *
+ * @deprecated use media-selection/overlay@sulumedia instead
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection-overlay/main.js]: Has the interface of the new media selection overlay significantly changed? Is it worth keeping  the old file?

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/main.js [hunk 5]
```diff
@@ -97,42 +97,14 @@ define(function() {
...
-                    }
-
-                    this.setData(data, false);
-                    this.removeItemById(itemId);
-                }.bind(this)
-            );
-
             this.sandbox.on('sulu.media-selection.' + this.options.instanceName + '.add-button-clicked', function() {
+                var items = _.map(this.getData().ids, function(id) {
+                    return {id: id};
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/main.js]: Why does this have to be an object?

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/overlay/main.js [hunk 6]
```diff
@@ -0,0 +1,564 @@
...
+            );
+        },
+
+        addFilesToDatagrid: function(files) {
+            for (var i = -1, length = files.length; ++i < length;) {
+                files[i].selected = true;
+            }
+
+            this.sandbox.emit('husky.datagrid.' + this.options.instanceName + '.records.add', files);
+            this.sandbox.emit('husky.data-navigation.' + this.options.instanceName + '.collections.reload');
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/overlay/main.js]: Why is a reload needed? Doesn't `records.add` also add the item visually?

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/main.js [hunk 7]
```diff
@@ -12,6 +12,9 @@ require.config({
         sulumedia: '../../sulumedia/js',
         sulumediacss: '../../sulumedia/css',
 
+        // if content will be registered later
+        sulucontentcss: '../../sulucontent/css',
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/main.js]: Why do we need the sulu content css here at all?

File: src/Sulu/Bundle/MediaBundle/Resources/public/scss/modules/single-media-selection.scss [hunk 8]
```diff
@@ -0,0 +1,153 @@
+.single-media-selection {
+    & {
+        height: 500px;
+
+        &, .slide, .overlay-footer {
+            width: 930px !important;
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/scss/modules/single-media-selection.scss]: Why are there so many `!important`s in this file? 😮

File: src/Sulu/Bundle/MediaBundle/Resources/translations/sulu/backend.de.xlf [hunk 9]
```diff
@@ -286,6 +286,38 @@
...
+                <target>Medien Link entfernen</target>
+            </trans-unit>
+            <trans-unit id="243d88cf2f64352w337js55369f111117" resname="media-selection.overlay.title">
+                <source>mediamedia-selection.overlay.title</source>
+                <target>Medien</target>
+            </trans-unit>
+            <trans-unit id="243d88cf2f64352w337js55369f111118" resname="media-selection.overlay.save">
+                <source>mediamedia-selection.overlay.save</source>
+                <target>Übernehmen</target>
+            </trans-unit>
```
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/translations/sulu/backend.de.xlf]: Use `sulu-media` as prefix for the media.

**Comment by danrot** [comment]
The media is only parsed on the website if there is a translation for the given language. Should that be the case, or should the fallback be applied as in the backend? /cc @chirimoya

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/ckeditor/media-link.js (hunk 3 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/ckeditor/media-link.js]: There is a "content" property, which gets the translated strings as the value.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/main.js (hunk 5 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/main.js]: That could be changed, but keep in mind that it is at the interface of an event-handler, that would be a BC break.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/overlay/main.js (hunk 6 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/overlay/main.js]: The second one is 'data-navigation' not 'datagrid'.

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/main.js (hunk 7 above)
Comment (marcelmoosbrugger) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/main.js]: I think @wachterjohannes needed it to load the css for the ck-editor extension.

**Comment by marcelmoosbrugger** [comment]
@danrot I made the media load in all languages. If the meta information is not available in the requested language, default meta get displayed.

File: src/Sulu/Bundle/ContentBundle/Resources/translations/sulu/backend.fr.xlf [hunk 10]
```diff
@@ -611,17 +611,17 @@
                 <source>content.changed-warning.description</source>
                 <target>Cette page a été changée par un autre utilisateur dans l’intervalle. En cas que cette page est sauvegardée, ces données se perdent.</target>
             </trans-unit>
-            <trans-unit id="26" resname="content.changed-warning.ok-button">
+            <trans-unit id="26" resname="content.changed-warning.ok-buttoyn">
```
Comment (danrot) [inline src/Sulu/Bundle/ContentBundle/Resources/translations/sulu/backend.fr.xlf]: `button`

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/main.js (hunk 5 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/components/media-selection/main.js]: ok, then leave it

File: src/Sulu/Bundle/MediaBundle/Resources/public/js/main.js (hunk 7 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/js/main.js]: I don't get it... Would it work if you just remove it? If possible I would like to minimize the dependencies between the different bundles...

File: src/Sulu/Bundle/MediaBundle/Resources/public/scss/modules/single-media-selection.scss (hunk 8 above)
Comment (danrot) [inline src/Sulu/Bundle/MediaBundle/Resources/public/scss/modules/single-media-selection.scss]: Okay, this only affects the `.single-media-selection`, which is currently only used in the ckeditor plugin... We should maybe still try to get it running without these, only add these important styles in the ckeditor context. Can you maybe add a comment for the moment, and create an issue for now?
